README.md
# Functions' Templates
This is the monorepo for templates that are used in various functions
## Templates in repo
1. OAuth Function Template


oauth-function-template/CHANGELOG.md
CHANGELOG
=============

Release 1.6.1
- Changes
  - Added fixes to work well with function-wrapper v1.2.5
  - Added changelog

Release 1.6.2
- Changes
  - Added fix to create properly new column (errorCode) in `OAuth Table`
  
Release 1.7.0
- Changes
  - Add optional headers and method to inner Authorization

Release 1.7.1
- Changes
  - Fixed setting up a token when removing credentials
  
Release 1.7.2
- Changes
  - Upgrade convict to 6.0.0


oauth-function-template/index.js
'use strict';

global.APP_ID = process.env.APP_ID || undefined;

const bootstrap = require('@dronedeploy/function-wrapper');
const config = require('./config.json');
const { createConfig } = require('./oauth-config');

const authorizationCode = require('./handlers/authorizationCode');
const clientCredentials = require('./handlers/clientCredentials');

exports.createOAuth = function(configuration) {
  const { authorizationCodeSettings, clientCredentialsSettings } = parseConfig(configuration);
  let paths = {};

  if (authorizationCodeSettings) {
    paths = authorizationCode.initHandler(createConfig(authorizationCodeSettings));
  }

  if (clientCredentialsSettings) {
    paths = Object.assign(paths, clientCredentials.initHandler(createConfig(clientCredentialsSettings)));
  }

  return handler(paths);
};

const parseConfig = function (configuration) {
  const newConfigFormat = configuration.clientCredentialsSettings || configuration.authorizationCodeSettings;
  if (newConfigFormat) {
    return {
      clientCredentialsSettings: configuration.clientCredentialsSettings,
      authorizationCodeSettings: configuration.authorizationCodeSettings,
    }
  }
  return {
    authorizationCodeSettings: configuration
  };
};

const handler = function(paths) {
  return (req, res) => {
    console.log(`Request received to endpoint: ${req.method} ${req.originalUrl}`);
    if (!global.APP_ID) {
      const msg = 'App ID not available, did you deploy using DroneDeploy-CLI?';
      console.error(msg);
      res.status(500).send(msg);
    }
    bootstrap(config, req, res, (err, ctx) => {
      if (err) {
        console.error(err, err.stack);
        console.warn('An error occurred during the bootstrapping process. A default response has been sent and code paths have been stopped.');
        return;
      }

      const pathHandler = paths[req.path];
      if (pathHandler) {
        pathHandler(req, res, ctx);
      } else {
        res.status(404).send('Not Found');
      }
    });
  }
};


oauth-function-template/oauth-config.js
const convict = require('convict');
const stringFormat = require('string-format');
const _ = require('lodash');

// Define a schema
const SCHEMA = {
  callbackUrl: {
    doc: "Callback url to be used during the OAuth flow",
    format: String,
    default: undefined
  },
  credentials: {
    client: {
      id: {
        doc: "Client ID value obtained from auth provider for use with APIs",
        format: String,
        sensitive: true,
        default: undefined // We can pull from ENV variables if possible
      },
      secret: {
        doc: "Client Secret value obtained from auth provider for use with APIs",
        format: String,
        sensitive: true,
        default: undefined // We can pull from ENV variables if possible
      }
    },
    auth: {
      authorizeHost: {
        doc: "Base host url for OAuth2 authorization",
        format: String,
        default: undefined
      },
      authorizePath: {
        doc: "Path on base authorizeHost domain for authorization",
        format: String,
        default: undefined
      },
      tokenHost: {
        doc: "Base host url for getting OAuth2 access tokens",
        format: String,
        default: undefined
      },
      tokenPath: {
        doc: "Path on base tokenHost domain for retrieving access tokens",
        format: String,
        default: undefined
      }
    },
    options: {
      useBasicAuthorizationHeader: {
        doc: "Whether or not the Basic Authorization header should be sent at the token request",
        format: "Boolean",
        default: undefined
      }
    }
  },
  authorizeUrl: {
    redirect_uri: {
      doc: "Fully qualified url for redirect following successful OAuth authorization",
      format: String,
      default: undefined
    },
    scope: {
      doc: "Authorization scope being requested from the OAuth provider",
      format: String,
      default: undefined
    }
  },
  innerAuthorization: {
    url: {
      doc: "Fully qualified url for inner authorization request made during refresh",
      format: String,
      default: undefined
    },
    removeCredentials: {
      doc: "Whether or not tokens should be removed in case of inner authorization failure",
      format: "Boolean",
      default: undefined
    },
    headers: {
      doc: "Optional headers used to request innerAuthorization.url",
      format: Object,
      default: undefined
    },
    method: {
      doc: "Optional http method used to request innerAuthorization.url",
      format: Object,
      default: undefined
    },
  },
  tokenConfig: {
    scope: {
      doc: "Authorization scope being requested from the OAuth provider",
      format: String,
      default: undefined
    }
  },
  credentialKeys: {
    clientId: {
      doc: "Client ID key to match from ENV variables",
      format: String,
      default: undefined
    },
    clientSecret: {
      doc: "Client Secret key to match from ENV variables",
      format: String,
      default: undefined
    },
  }
};

module.exports.createConfig = (configuration) => {
  const config = convict(SCHEMA);
  config.load(configuration);
  let clientIdKey = config.has('credentialKeys.clientId') ? config.get('credentialKeys.clientId') : 'CLIENT_ID';
  if (!(_.get(process.env, clientIdKey))) {
    throw new Error(`${clientIdKey} environment variable not set`);
  }
  let clientSecretKey = config.has('credentialKeys.clientSecret') ? config.get('credentialKeys.clientSecret') : 'CLIENT_SECRET';
  if (!(_.get(process.env, clientSecretKey))) {
    throw new Error(`${clientSecretKey} environment variable not set`);
  }

  config.set('credentials.client.id', _.get(process.env, clientIdKey));
  config.set('credentials.client.secret', _.get(process.env, clientSecretKey));

  if (config.get('callbackUrl')) {
    // This formats the callback url dynamically based on the deployed function
    // name. For example the function name of fn-123456789 would be inserted into
    // the templated spot in the callback url https://dronedeployfunctions.com/{}/route
    config.set('callbackUrl', stringFormat(config.get('callbackUrl'), process.env.FUNCTION_NAME));
    config.set('authorizeUrl.redirect_uri', config.get('callbackUrl'));
  }

  return config;
};



oauth-function-template/datastore/table.js
const queries = require('./queries');

// Returns tableId
const _oauthTableSetup = (ctx) => {
  return _createOAuthTableIfNotExists(ctx)
    .catch((err) => {
      throw new Error(err);
    });
};

const _createOAuthTableIfNotExists = (ctx) => {
  return ctx.graphql.query(queries.findTableQuery)
      .then((result) => {
        if (result.errors && result.data.node.table === null) {
          console.log('Creating OAuth table as it does not exist');
          return _createOAuthTable(ctx)
            .then((tableIdResult) => {
              console.log('Created OAuth table without columns');
              return createOAuthTableColumns(ctx, tableIdResult);
            });
        }

        console.log('OAuth table exists - no need to create one.');
        return result.data.node.table.id;
      });
};

const _createOAuthTable = (ctx) => {
  const createInput = {
    input: {
      applicationId: `Application:${global.APP_ID}`,
      name: queries.tableName,
      description: "Contains token data for OAuth"
    }
  };
  return ctx.graphql.query(queries.createTableQuery, createInput)
    .then((result) => {
      if (result.errors) {
        return Promise.reject(result.errors[0]);
      }

      return result.data.createTable.table.id;
    });
};

const resultContainsError = (result) => {
  return !!result.errors;
};

const getMissingColumns = (columnsSubset) => {
  return queries.tableColumns.filter((column) => {
    return !columnsSubset.includes(column.name);
  });
};

const getOAuthTableColumnsToCreate = (ctx, tableId) => {
  return ctx.graphql.query(queries.findColumnsQuery)
    .then((result) => {
      var columnsResult = result.data.node.table.columns;
      switch (columnsResult.length) {
        case 0:
          return queries.tableColumns;
        case 4:
          return [];
        default:
          return getMissingColumns(columnsResult);
      }
    })
};

const createOAuthTableColumns = (ctx, tableId) => {
  console.log('Creating OAuth table columns');
  return getOAuthTableColumnsToCreate(ctx, tableId)
    .then((columns) => {
      // This really shouldn't ever happen, but if it does
      // we don't want to error out trying to create columns
      // that already exist
      if (columns.length === 0) {
        return tableId;
      }

      var tableColumnQueries = columns.map((columnData) => {
        columnData.input.tableId = tableId;
        return ctx.graphql.query(queries.createTableColumnQuery, columnData);
      });
    
      return Promise.all(tableColumnQueries)
        .then((results) => {
          if (results.some(resultContainsError)) {
            return Promise.reject('Error creating OAuth table columns');
          }
          console.log('OAuth Table columns created');
          return tableId;
        });
    });
};

module.exports = {
  setupOAuthTable: _oauthTableSetup
};


oauth-function-template/datastore/queries.js
const TABLE_NAME = "OAuth Table";

const CREATE_TABLE_QUERY = `mutation CreateTable($input: CreateTableInput!) {
  createTable(input: $input) {
    table {
      id
      application {
        id
      }
      name
      description
    }
  }
}`;

const CREATE_TABLE_COLUMN_QUERY = `mutation CreateTableColumn($input: CreateTableColumnInput!) {
  createTableColumn(input: $input) {
    tableColumn {
      id
      name
      description
      ... on NumberColumn {
        type
      }
      ... on TextColumn {
        length
      }
    }
  }
}`;

const FIND_TABLE_QUERY = `{
  node(id: "Application:${global.APP_ID}") {
    ... on Application {
      table(name: "${TABLE_NAME}") {
        id
        name
      }
    }
  }
}`;

const FIND_COLUMNS_QUERY = `{
  node(id: "Application:${global.APP_ID}") {
    ... on Application {
      table(name: "${TABLE_NAME}") {
        name
        columns {
          name
        }
      }
    }
  }
}`;

let ACCESS_TOKEN_COLUMN = {
  input: {
    columnType: "TEXT",
    name: "accessToken",
    textLength: 4096,
    textEncrypted: true,
    description: "Holds the current access token"
  }
};

let TOKEN_EXPIRATION_COLUMN = {
  input: {
    columnType: "DATETIME",
    name: "access_expires_at",
    description: "Holds the date and time at which the current access token expires"
  }
};

let REFRESH_TOKEN_COLUMN = {
  input: {
    columnType: "TEXT",
    name: "refreshToken",
    textLength: 4096,
    textEncrypted: true,
    description: "Holds the current refresh token"
  }
};

let ERROR_CODE_COLUMN = {
  input: {
    columnType: "NUMBER",
    name: "errorCode",
    description: "Holds the latest error code"
  }
};

let TABLE_COLUMNS = [
  ACCESS_TOKEN_COLUMN,
  TOKEN_EXPIRATION_COLUMN,
  REFRESH_TOKEN_COLUMN,
  ERROR_CODE_COLUMN,
];

module.exports = {
  createTableColumnQuery: CREATE_TABLE_COLUMN_QUERY,
  createTableQuery: CREATE_TABLE_QUERY,
  findColumnsQuery: FIND_COLUMNS_QUERY,
  findTableQuery: FIND_TABLE_QUERY,
  tableColumns: TABLE_COLUMNS,
  tableName: TABLE_NAME
};


oauth-function-template/handlers/clientCredentials.js
const oauth2 = require('simple-oauth2');
const tableUtils = require('../datastore/table');


exports.initHandler = function(config) {
  const handlerClass = new ClientCredentialsHandler(config);
  return handlerClass.getRoutes();
};


class ClientCredentialsHandler {
  constructor(config) {
    this.datastoreExternalId = 'clientCredentialsToken';
    this.expirationWindowInSeconds = 300;
    this.tokenConfig = config.get('tokenConfig');
    this.oauth2 = oauth2.create(config.get('credentials'));
  }

  getRoutes() {
    return {
      '/refreshClientCredentials': this.refreshClientCredentialsHandler.bind(this),
    }
  }

  refreshClientCredentialsHandler(req, res, ctx) {
    return tableUtils.setupOAuthTable(ctx)
      .then((tableId) => ctx.datastore.table(tableId))
      .then((table) => {
        return table.getRowByExternalId(this.datastoreExternalId)
          .then((row) => {
            const needRefresh = !row.ok || this.doesTokenNeedRefresh(row.data);
            if (needRefresh) {
              console.log('Client credentials token needs to be refreshed - refreshing.');
              return this.refreshClientCredentials(this.tokenConfig)
                .then((tokenObj) => this.storeClientCredentialsToken(table, tokenObj))
                .then((result) => {
                  console.log('Client credentials token refreshed.');
                  return result;
                });
            }
            console.log('Client credentials are fresh - no need to refresh.');
            return Promise.resolve();
          })
          .then(() => res.status(200).send());
      })
      .catch((error) => {
        const msg = 'Refresh client credentials token failed';
        console.error(msg, error);
        return res.status(500).send(msg);
      });
  }

  refreshClientCredentials(tokenConfig) {
    return this.oauth2.clientCredentials.getToken(tokenConfig)
      .then((result) => this.oauth2.accessToken.create(result))
      .then((accessToken) => ({
        accessToken: accessToken.token.access_token,
        access_expires_at: accessToken.token.expires_at,
        refreshToken: ''
      }));
  }

  doesTokenNeedRefresh(token) {
    const isTokenIncomplete = !token.accessToken || !token.access_expires_at;
    const expirationTimeInSeconds = new Date(token.access_expires_at).getTime() / 1000;
    const nowInSeconds = new Date().getTime() / 1000;
    const shouldRefresh = nowInSeconds >= (expirationTimeInSeconds - this.expirationWindowInSeconds);
    return isTokenIncomplete || shouldRefresh;
  }

  storeClientCredentialsToken(table, token) {
    return table.upsertRow(this.datastoreExternalId, token)
      .then((rowData) => {
        if (!rowData.ok) {
          throw new Error(`An error occured while updating token data in Datastore: ${JSON.stringify(rowData.errors)}`);
        }
      });
  }
}


oauth-function-template/handlers/authorizationCode.js
let oauth2 = require('simple-oauth2');
const tableUtils = require('../datastore/table');
let fetch = require('node-fetch');
const _ = require('lodash');

const provider = {};
const SERVICE_ACCOUNT_EXTERNAL_KEY = 'serviceAccount';

const ERROR_CODES = {
    INNER_AUTHORIZATION_FAILED: { code: 0, message: 'Inner Authorization Failed' },
};

exports.initHandler = function(config) {
  provider.callbackUrl = config.get('callbackUrl');
  provider.credentials = config.get('credentials');
  provider.authorizeUrl = config.get('authorizeUrl');
  provider.innerAuthorization = config.get('innerAuthorization');
  oauth2 = oauth2.create(provider.credentials);

  return {
    '/refresh': refreshHandler,
    '/auth': oauth2InitHandler,
    '/auth/callback': oauth2CallbackHandler,
    '/token': storeTokenHandler,
    '/logout': logoutHandler,
  }
};

const refreshHandler = (req, res, ctx) => {
  return tableUtils.setupOAuthTable(ctx)
    .then((tableId) => {
      const accessTokensTable = ctx.datastore.table(tableId);
      const storageTokenInfo = getStorageTokenInfo(req, ctx);

      return accessTokensTable.getRowByExternalId(storageTokenInfo.externalId)
        .then((result) => {
          if (!result.ok) {
            if (couldNotFindRowData(result)) {
              console.log('User is not signed in - queried OAuth table row does not exist');
              return res.status(401).send(packageError(result));
            }
            console.error(`An error occurred during oauth table row retrieval: ${JSON.stringify(result.errors)}`);
            return res.status(500).send(packageError(result));
          }
          const url = _.get(provider, 'innerAuthorization.url') || _.get(provider, 'innerAuthorizationUrl');
          if (url && !!result.data.accessToken) {
              const headers = _.get(provider, 'innerAuthorization.headers');
              const method = _.get(provider, 'innerAuthorization.method');
              return getInnerAuthorizationResponse(url, result.data.accessToken, headers, method)
                  .then((isOkStatus) => {
                      if (!isOkStatus) {
                          console.warn(
                            'The access token is invalid - removing access token in a related datastore row'
                          );
                          const removeCredentials = _.get(provider, 'innerAuthorization.removeCredentials', true);
                          const replaceableToken = removeCredentials ?
                              emptyTokenWithErrorCode(ERROR_CODES.INNER_AUTHORIZATION_FAILED.code) :
                              oldTokenWithErrorCode(result.data, ERROR_CODES.INNER_AUTHORIZATION_FAILED.code);
                          accessTokensTable.editRow(storageTokenInfo.externalId, replaceableToken);
                          return res.status(204).send(ERROR_CODES.INNER_AUTHORIZATION_FAILED);
                      } else {
                        console.log('Access token is valid - proceeding.');
                        return getStandardAuthorizationResponse(res, result, accessTokensTable, storageTokenInfo);
                      }
                  });
          }
          return getStandardAuthorizationResponse(res, result, accessTokensTable, storageTokenInfo);
        })
    })
};

const getInnerAuthorizationResponse = (url, accessToken, headers = {}, method = 'GET') => {
    let options = {
        method,
        headers: {
            'Authorization': `Bearer ${accessToken}`,
            'Content-Type': 'application/json',
            ...headers,
        }
    };
    console.info(`Making a check if the access token is still valid by doing a request: ${method} ${url}`);
    return fetch(url, options)
        .then((res) => {
            return res.ok;
        })
        .catch((error) => {
                console.error('An error during checking if access token is still valid occured.', error);
                return false;
        });
};

const getStandardAuthorizationResponse = (res, result, accessTokensTable, storageTokenInfo) => {
    if (isAccessTokenValidForever(result.data)) {
        console.log('Related access token is valid forever - no refresh required.');
        return res.status(200).send();
    }
    console.log(`Related access token expires at: ${result.data.access_expires_at}`);

    if (isEmptyToken(result.data)) {
        console.log('User needs to sign in - access token is empty in a related datastore row');
        return res.status(204).send();
    }

    // This create call handles keeping tokens for us in this accessToken
    // object, but in reality, we would want to save in datastore
    const tokenData = {
        access_token: result.data.accessToken,
        expires_at: result.data.access_expires_at,
        refresh_token: result.data.refreshToken
    };
    const accessTokenObj = oauth2.accessToken.create(tokenData);

    // We preemptively refresh the token to avoid sending a token back
    // to the client that may expire very soon
    if (doesTokenNeedRefresh(accessTokenObj.token)) {
        console.log("Refreshing token as it is expired or will be expired soon.");
        return accessTokenObj.refresh()
            .then((refreshResult) => {
                console.log("Refresh success");
                return storeTokenData(accessTokensTable, storageTokenInfo, refreshResult.token, res);
            })
            .catch((err) => {
                console.error('Access token refresh failed', err);
                return res.status(401).send(packageError('The authorization code/refresh token is expired or invalid/redirect_uri must have the same value as in the authorization request.'));
            });
    }

    console.log(`No refresh required as the access token is not going to expire soon.`);
    console.log(`${storageTokenInfo.returnTokenBack ? '' : 'Not'} returning access token in the response`);
    return res.status(200).send(storageTokenInfo.returnTokenBack ? tokenData.access_token : '');
};

const getStorageTokenInfo = (req, ctx) => {
  const isServiceAccount = req.query.service_account === 'true';
  console.log(
    isServiceAccount ? 'Received request is for Service account' : 'Received request is for non-service account'
  );
  return {
    externalId: isServiceAccount ? SERVICE_ACCOUNT_EXTERNAL_KEY : ctx.token.username,
    returnTokenBack: !isServiceAccount,
  };
};

const couldNotFindRowData = (result) => {
  return !result.errors || result.errors[0].message.includes('Could not find data');
};

const packageError = (err) => {
  return { error: err };
};

const isEmptyToken = (data) => {
  const isUndefined = (!data.accessToken || !data.refreshToken);
  const isEmpty = (data.accessToken === "" || data.refreshToken === "");
  return isUndefined || isEmpty;
};

const isAccessTokenValidForever = ({ refreshToken, accessToken, access_expires_at }) => {
  return !refreshToken && !!accessToken && access_expires_at === '1970-01-01T00:00:00.000Z';
};

const doesTokenNeedRefresh = (token) => {
  // Provide a window of time before the actual expiration to
  // refresh the token
  const EXPIRATION_WINDOW_IN_SECONDS = 300;

  const expirationTimeInSeconds = token.expires_at.getTime() / 1000;
  const expirationWindowStart = expirationTimeInSeconds - EXPIRATION_WINDOW_IN_SECONDS;

  // If the start of the window has passed, refresh the token
  const nowInSeconds = (new Date()).getTime() / 1000;
  return nowInSeconds >= expirationWindowStart;
};

// Stores the token in the datastore and return it to the client if successful
const storeTokenData = (table, storageTokenInfo, tokenData, res) => {
  // Some tokens may not have an 'expires_at' property, so we will
  // calculate it anyway based on the 'expires_in' value
  const accessTokenObj = oauth2.accessToken.create(tokenData);
  return table.upsertRow(storageTokenInfo.externalId, {
    accessToken: accessTokenObj.token.access_token,
    access_expires_at: accessTokenObj.token.expires_at,
    errorCode: null,
    refreshToken: accessTokenObj.token.refresh_token}
  ).then((rowData) => {
    if (!rowData.ok) {
      // Problem storing the access token which will
      // impact potential future api calls - send error
      throw new Error(JSON.stringify(rowData.errors[0]));
    }

    console.log(storageTokenInfo.returnTokenBack ? 'Returning new token' : 'Not returning new token')
    return res.status(200).send(storageTokenInfo.returnTokenBack ? accessTokenObj.token : '');
  });
};

// Initializes the OAuth2 flow - successful authorization results in redirect to callback export
const oauth2InitHandler = (req, res, ctx) => {
  provider.authorizeUrl.redirect_uri = `${provider.callbackUrl}`;
  const authorizationUri = oauth2.authorizationCode.authorizeURL(provider.authorizeUrl);
  res.send(authorizationUri);
};

// Callback handler to take the auth code from first OAuth2 step and get the tokens
const oauth2CallbackHandler = (req, res, ctx) => {
  const tokenConfig = {
    code: req.query.code,
    redirect_uri: provider.callbackUrl
  };

  return oauth2.authorizationCode.getToken(tokenConfig)
    .then((result) => {
      console.log('Successfuly produced oauth token from authorization code - returning callback html');
      return res.status(200).send(generateCallbackHtml(oauth2.accessToken.create(result).token));
    })
    .catch((error) => {
      console.log('Failure on producing oauth token from authorization code - returning error html ', error.message);
      return res.status(500).send(createErrorHtml(error.message));
    });
};

const storeTokenHandler = (req, res, ctx) => {
  // Make sure this is called with the proper method
  if (req.method !== 'POST' && req.method !== 'PUT') {
    console.warn(`Invalid http method: ${req.method} - returning 400`);
    return res.status(400).send(packageError('Invalid request method - please use POST or PUT'));
  }

  if (!req.body) {
    const message = 'Missing request body';
    console.warn(message);
    return res.status(400).send(packageError(message));
  }

  const parsed = JSON.parse(req.body);

  // Make sure user has passed correct data parameter
  if (!parsed.token) {
    const message = 'Missing token body in request';
    console.warn(message);
    return res.status(400).send(packageError(message));
  }

  const storageTokenInfo = getStorageTokenInfo(req, ctx);

  // Get our oauth table and store the token data
  return tableUtils.setupOAuthTable(ctx)
    .then((tableId) => {
      const accessTokensTable = ctx.datastore.table(tableId);

      // we store the access token data by associating
      // it with the user on the function jwt auth token
      console.log('Saving token in datastore');
      return storeTokenData(accessTokensTable, storageTokenInfo, parsed.token, res);
    })
    .catch((error) => {
      console.log('Error storing Access Token', error.message);
      return res.status(500).send(packageError(error.message));
    });
};

const createErrorHtml = (err) => {
  // Nest in an error field for consistency at the client-level
  // and so if err is a string, we create an object for the callback html
  return generateCallbackHtml(packageError(err));
};

const generateCallbackHtml = (token) => {
  token = JSON.stringify(token);
  return '<!DOCTYPE html>'+
  '<html lang="en">'+
  '<head>'+
  '    <meta charset="UTF-8">'+
  '    <meta name="viewport" content="width=device-width, initial-scale=1.0">'+
  '    <meta http-equiv="X-UA-Compatible" content="ie=edge">'+
  '    <title>Document</title>'+
  '</head>'+
  '<body>'+
  '    <script>'+
  '    window.onload = function() {'+
  `        window.opener.postMessage(\'${token}\', \'*\')`+
  '    }'+
  '    </script>'+
  '</body>'+
  '</html>';
};

const logoutHandler = (req, res, ctx) => {
  return tableUtils.setupOAuthTable(ctx)
    .then((tableId) => {
      const accessTokensTable = ctx.datastore.table(tableId);
      const storageTokenInfo = getStorageTokenInfo(req, ctx);

      console.log('Signing out user - setting related OAuth table row to an empty token');
      return accessTokensTable.editRow(storageTokenInfo.externalId, emptyTokenWithErrorCode())
        .then((result) => {
          if (!result.ok) {
            console.error(`Error on setting empty token: ${JSON.stringify(result.errors)} - returning error html`);
            return res.status(500).send(createErrorHtml(result.errors[0]));
          }
          console.log('Sign out successful');
          return res.status(200).send(generateCallbackHtml({}));
        });
    });
};

// Blank tokens and current date (needed for date column validation)
// for use with logout
const emptyTokenWithErrorCode = (errorCode = null) => ({
    accessToken: "",
    access_expires_at: new Date().toISOString(),
    errorCode,
    refreshToken: "",
});

const oldTokenWithErrorCode = (oldValue, errorCode = null) => ({
    ...oldValue,
    errorCode,
});
